File: src/java/eu/sig/training/ch02/Level.java
package eu.sig.training.ch02;

import java.util.List;

public class Level {
    private boolean inProgress;
    private final List<LevelObserver> observers;

    Level(List<LevelObserver> observers) {
        this.observers = observers;
    }

    // tag::start[]
    public void start() {
        if (inProgress) {
            return;
        }
        inProgress = true;
        // Update observers if player died:
        if (!isAnyPlayerAlive()) {
            for (LevelObserver o : observers) {
                o.levelLost();
            }
        }
        // Update observers if all pellets eaten:
        if (remainingPellets() == 0) {
            for (LevelObserver o : observers) {
                o.levelWon();
            }
        }
    }
    // end::start[]

    @SuppressWarnings("unused")
    // tag::updateObservers[]
    private void updateObservers() {
        // Update observers if player died:
        if (!isAnyPlayerAlive()) {
            for (LevelObserver o : observers) {
                o.levelLost();
            }
        }
        // Update observers if all pellets eaten:
        if (remainingPellets() == 0) {
            for (LevelObserver o : observers) {
                o.levelWon();
            }
        }
    }
    // end::updateObservers[]

    @SuppressWarnings("unused")
    // tag::updateObserversPlayerDied[]
    private void updateObserversPlayerDied() {
        if (!isAnyPlayerAlive()) {
            for (LevelObserver o : observers) {
                o.levelLost();
            }
        }
    }

    // end::updateObserversPlayerDied[]

    @SuppressWarnings("unused")
    // tag::updateObserversPelletsEaten[]
    private void updateObserversPelletsEaten() {
        if (remainingPellets() == 0) {
            for (LevelObserver o : observers) {
                o.levelWon();
            }
        }
    }
    // end::updateObserversPelletsEaten[]

    private int remainingPellets() {
        return 0;
    }

    private boolean isAnyPlayerAlive() {
        return false;
    }
}

class LevelObserver {
    public void levelLost() {}
    public void levelWon() {}
}

File: src/csharp/eu/sig/training/ch02/Level.cs
using System.Collections.Generic;

namespace eu.sig.training.ch02
{

    public class Level
    {
        private bool inProgress;
        private readonly IList<LevelObserver> observers;

        Level(IList<LevelObserver> observers)
        {
            this.observers = observers;
        }

        // tag::start[]
        public void Start()
        {
            if (inProgress)
            {
                return;
            }
            inProgress = true;
            // Update observers if player died:
            if (!IsAnyPlayerAlive())
            {
                foreach (LevelObserver o in observers)
                {
                    o.LevelLost();
                }
            }
            // Update observers if all pellets eaten:
            if (RemainingPellets() == 0)
            {
                foreach (LevelObserver o in observers)
                {
                    o.LevelWon();
                }
            }
        }
        // end::start[]

        // tag::updateObservers[]
        private void UpdateObservers()
        {
            // Update observers if player died:
            if (!IsAnyPlayerAlive())
            {
                foreach (LevelObserver o in observers)
                {
                    o.LevelLost();
                }
            }
            // Update observers if all pellets eaten:
            if (RemainingPellets() == 0)
            {
                foreach (LevelObserver o in observers)
                {
                    o.LevelWon();
                }
            }
        }
        // end::updateObservers[]

        // tag::updateObserversPlayerDied[]
        private void UpdateObserversPlayerDied()
        {
            if (!IsAnyPlayerAlive())
            {
                foreach (LevelObserver o in observers)
                {
                    o.LevelLost();
                }
            }
        }

        // end::updateObserversPlayerDied[]

        // tag::updateObserversPelletsEaten[]
        private void UpdateObserversPelletsEaten()
        {
            if (RemainingPellets() == 0)
            {
                foreach (LevelObserver o in observers)
                {
                    o.LevelWon();
                }
            }
        }
        // end::updateObserversPelletsEaten[]

        private int RemainingPellets()
        {
            return 0;
        }

        private bool IsAnyPlayerAlive()
        {
            return false;
        }
    }

    class LevelObserver
    {
        public void LevelLost() { }
        public void LevelWon() { }
    }

}
